### PR TITLE
Fix #7944: Demolishing locks built on rivers didn't always restore the river

### LINKED ISSUE #7944: Demolishing a lock built on a river with rocky land does not restore the river
## Version of OpenTTD
1.10.0-beta2
## Expected result
River is restored at the middle tile
## Actual result
River is not restored at the middle tile
## Steps to reproduce
Use NewGRF OpenGFX+ Landscape 1.1.2
Place Rocky land object on sloped river.
Build a lock.
Demolish lock.

### PR DESCRIPTION
**Pull request by SamuXarick**

### CODE REVIEW

**Comment by LordAro** [comment]
Having tested it, it's not immediately clear why `IsWaterTile` does not work for the sloped water, but does work for the flat water. Should `HasTileWaterGround` be used for the ends of the lock as well?

**Comment by SamuXarick** [comment]
It works for the flat water because...

The lower and upper tiles of the lock are getting WaterClass after the clear command in the case of object on river (rocky land). In Test mode, the WaterClass becomes WATER_CLASS_CANAL by failing IsWaterTile check, because the clear command was also in test mode. But in Exec mode, the tile resulting by the clear command is a river tile, without the rocks, WaterClass then is WATER_CLASS_RIVER.

`MakeLock(tile, _current_company, dir, wc_lower, wc_upper, wc_middle);` is only called in DC_EXEC mode, garanteeing the correct wc_upper/wc_lower.

Then when removing the lock, `MakeWaterKeepingClass` will ensure rivers are restored, by looking at the WaterClass of those 2 tiles.

In short, DC_TEST and DC_EXEC are being toyed around with the WaterClass for the upper/lower tiles. The middle tile is not. WaterClass is being checked before the clear command. For a river with rocks on it, IsTileWater test false, thus WaterClass for middle tile is WATER_CLASS_CANAL, but for HasTileWaterGround it returns WATER_CLASS_RIVER.

**Review by LordAro — DISMISSED** [review]

**Comment by LordAro** [comment]
@SamuXarick This needs a rebase due to commit-checker changes, can you do?

**Review by LordAro — APPROVED** [review]
